pkg/segment/writer/agiletree.go

```diff
@@ -279,6 +280,149 @@ func (stb *StarTreeBuilder) insertIntoTree(node *Node, colVals []uint32, recNum
 	}
 }
 
+func (stb *StarTreeBuilder) updateAggVals(node *Node, nodeToMerge *Node) error {
+	if nodeToMerge.aggValues == nil {
+		return nil
+	}
+	if node.aggValues == nil {
+		node.aggValues = make([]utils.CValueEnclosure, len(stb.mColNames)*TotalMeasFns)
+	}
+
+	var err error
+	for mcNum := range stb.mColNames {
+		midx := mcNum * TotalMeasFns
+		agvidx := midx + MeasFnMinIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Min)
+		if err != nil {
+			log.Errorf("updateAggVals: error in min err:%v", err)
+			return err
+		}
+
+		agvidx = midx + MeasFnMaxIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Max)
+		if err != nil {
+			log.Errorf("updateAggVals: error in max err:%v", err)
+			return err
+		}
+
+		agvidx = midx + MeasFnSumIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Sum)
+		if err != nil {
+			log.Errorf("updateAggVals: error in sum err:%v", err)
+			return err
+		}
+
+		agvidx = midx + MeasFnCountIdx
+		node.aggValues[agvidx], err = utils.Reduce(node.aggValues[agvidx], nodeToMerge.aggValues[agvidx], utils.Count)
+		if err != nil {
+			log.Errorf("updateAggVals: error in count err:%v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (stb *StarTreeBuilder) cleanupCommonChildren(currNode *Node) error {
```

Comment:
So this recursively merges nodes lower in the tree where the nodes are the same now that a level somewhere above was deleted, right?

Reply:
it properly links the children of the common nodes to the first node along with the aggs. Then, it recursively calls this for their own children and so on. While returning it links the node with the first child and cleans up accumulated nodes i.e. `commonChildren`.

Reply:
Ok nice. When the PR is marked Ready for Review it would probably be good to rename this function to make it more obvious. I think there were a few other things that could be good to rename, but I didn't note them during my review since the PR is Draft

Reply:
addressed in the latest commit.